### PR TITLE
Re-work gherkin parsing and scenario outlines.

### PR DESCRIPTION
**Pull request by techcobweb**
Signed-off-by: Mike Cobbett <77053+techcobweb@users.noreply.github.com>

# Why ?
See [Gherkin null pointer exception when using Examples table #1858](https://github.com/galasa-dev/projectmanagement/issues/1858)

# Tasks
- [x] Create unit test framework around the parsing of the gherkin feature
- [x] Add unit tests
- [x] Cope with bad URLs
- [x] Complete re-work of the parsing of the feature file text.
- [x] Error if `Examples:` used in a normal `Scenario:`
- [x] Error if `Examples:` is NOT used in a `Scenario outline:`
- [x] Single-line comments starting with `#` are ignored.
- [x] Re-write the parser, and the transform to get from an AST to the run-time structure.
- [ ] Explain what happens to comments
- [x] Test that it does what it should do on a real system

### CODE REVIEW

**Comment by galasa-team** [comment]
Build successful

**Comment by galasa-team** [comment]
Build successful

**Review by hobbit1983 — APPROVED** [review]

**Review by eamansour — APPROVED** [review]
Just a few trivial suggestions, but approved

File: galasa-parent/dev.galasa.framework/src/main/java/dev/galasa/framework/spi/language/gherkin/GherkinTest.java [hunk 1]
```diff
@@ -55,99 +51,128 @@ public class GherkinTest {
...
-                GherkinMethod currentMethod = null;
-                Section currentSection = Section.FEATURE;
-                boolean exampleHeaderLineProcessed = false;
-
-                for(String line : lines) {
-                    line = line.trim();
-                    if(line.isEmpty()) {
-                        continue;
-                    }
+        List<String> lines = getGhekinFeatureTextLines(run,fileReader);
```
[the remark this answers is not in the document]
Comment (techcobweb) [reply galasa-parent/dev.galasa.framework/src/main/java/dev/galasa/framework/spi/language/gherkin/GherkinTest.java:73]: Fixed.

File: galasa-parent/dev.galasa.framework/src/main/java/dev/galasa/framework/spi/language/gherkin/GherkinTest.java [hunk 2]
```diff
@@ -55,99 +51,128 @@ public class GherkinTest {
...
+                case EXAMPLE:
+                    if(exampleHeaderLineProcessed) {
+                        // We've already processed the header line.
+                        variables.processDataLine(line);
+                    } else {
+                        // Not yet processed the header line.
+                        variables.processHeaderLine(line);
+                        exampleHeaderLineProcessed = true;
+                    }
+
```
[the remark this answers is not in the document]
Comment (techcobweb) [reply galasa-parent/dev.galasa.framework/src/main/java/dev/galasa/framework/spi/language/gherkin/GherkinTest.java:140]: Yep. Added break.

File: galasa-parent/dev.galasa.framework/src/test/java/dev/galasa/framework/spi/language/gherkin/MockRun.java [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+    }
+
+    @Override
+    public String getGherkin() {
+        return gherkinFeatureText;
+    }
+
+    public void setGherkin(String featureText) {
+        this.gherkinFeatureText = featureText;
+    }
```
[the remark this answers is not in the document]
Comment (techcobweb) [reply galasa-parent/dev.galasa.framework/src/test/java/dev/galasa/framework/spi/language/gherkin/MockRun.java:148]: Ok. Don't see why not. We should add this to coding guidelines.

**Comment by galasa-team** [comment]
Build failed, see http://localhost:8001/api/v1/namespaces/tekton-pipelines/services/tekton-dashboard:http/proxy/#/namespaces/galasa-build/pipelineruns/repo-framework-pr-544-br6q2 for details. If you are unable to do so, please contact a member of the Galasa team.

**Comment by galasa-team** [comment]
Build failed, see http://localhost:8001/api/v1/namespaces/tekton-pipelines/services/tekton-dashboard:http/proxy/#/namespaces/galasa-build/pipelineruns/repo-framework-pr-544-shw9x for details. If you are unable to do so, please contact a member of the Galasa team.

**Comment by galasa-team** [comment]
Build successful

**Comment by galasa-team** [comment]
Build successful

**Comment by galasa-team** [comment]
Build successful

**Comment by galasa-team** [comment]
Build successful

**Comment by galasa-team** [comment]
Build successful

File: docs/gherkin.md [hunk 4]
```diff
@@ -0,0 +1,3 @@
+
+
+move terminal( \\w+)? cursor to field \"(.*)\"
```
Comment (techcobweb) [inline docs/gherkin.md:3]: This file needs deleting

**Comment by galasa-team** [comment]
Build successful

**Comment by galasa-team** [comment]
Build successful

**Pull request merged**